### PR TITLE
LPS-188397 | Display errors for Copy To size limits

### PR DESCRIPTION
**Pull request by jaimelr10**
**Description:** Added errors when limits are exceeded trying to copy a folder or file 

**Screenshots**
When Site or Instance limits are exceeded
image

When System limit is exceeded
image

**Jira Ticket:** https://liferay.atlassian.net/browse/LPS-188397

### CODE REVIEW

**Comment by liferay-lima** [comment]
ci:test:sf

**Comment by liferay-lima** [comment]
ci:test:relevant

File: portal-kernel/src/com/liferay/portal/kernel/service/persistence/impl/BasePersistenceImpl.java [hunk 1]
```diff
@@ -124,6 +124,8 @@ public class BasePersistenceImpl<T extends BaseModel<T>>
 
 	public static final String COUNT_COLUMN_NAME = "COUNT_VALUE";
 
+	public static final String SUM_COLUMN_NAME = "SUM_VALUE";
```
Comment (jaimelr10) [inline portal-kernel/src/com/liferay/portal/kernel/service/persistence/impl/BasePersistenceImpl.java:127]: This should be removed

Comment (AliciaGarciaGarcia) [inline portal-kernel/src/com/liferay/portal/kernel/service/persistence/impl/BasePersistenceImpl.java:127]: not needed

File: portal-impl/src/com/liferay/portlet/documentlibrary/service/impl/DLFolderLocalServiceImpl.java [hunk 2]
```diff
@@ -1425,6 +1487,36 @@ protected void validateFolderName(String folderName)
 		}
 	}
 
+	private long _getDLFolderSize(
```
Comment (AliciaGarciaGarcia) [inline portal-impl/src/com/liferay/portlet/documentlibrary/service/impl/DLFolderLocalServiceImpl.java:1490]: worth a public method, and use it on a common method for "canCopy"

File: portal-impl/src/com/liferay/portlet/documentlibrary/service/impl/DLFileEntryLocalServiceImpl.java [hunk 3]
```diff
@@ -366,6 +369,26 @@ public DLFileVersion cancelCheckOut(long userId, long fileEntryId)
 		return dlFileVersion;
 	}
 
+	@Override
+	public boolean canCopyFileEntry(
```
Comment (AliciaGarciaGarcia) [inline portal-impl/src/com/liferay/portlet/documentlibrary/service/impl/DLFileEntryLocalServiceImpl.java:373]: canCopyFile Entry And can extract a method because is the same behavior than with folder

File: portal-kernel/src/com/liferay/document/library/kernel/service/DLFileEntryLocalService.java [hunk 4]
```diff
@@ -55,6 +57,8 @@
 import java.util.List;
 import java.util.Map;
 
+import javax.portlet.RenderRequest;
```
Comment (AliciaGarciaGarcia) [inline portal-kernel/src/com/liferay/document/library/kernel/service/DLFileEntryLocalService.java:60]: separate the buildService on independent commit

File: portal-impl/src/com/liferay/portlet/documentlibrary/service/impl/DLFileEntryLocalServiceImpl.java [hunk 5]
```diff
@@ -1079,6 +1102,40 @@ public DLFileEntry fetchFileEntryByName(
...
+	@Override
+	public Object[] getCopyFailInfo(
+		FileEntry fileEntry, long systemMaxSizeToCopy,
+		long companyMaxSizeToCopy, long groupMaxSizeToCopy, Portal portal,
+		RenderRequest renderRequest) {
+
+		long dlFileEntrySize = fileEntry.getSize();
+
+		if ((dlFileEntrySize > groupMaxSizeToCopy) &&
+			(systemMaxSizeToCopy != 0)) {
```
Comment (AliciaGarciaGarcia) [inline portal-impl/src/com/liferay/portlet/documentlibrary/service/impl/DLFileEntryLocalServiceImpl.java:1114]: should be group?

**Review by AliciaGarciaGarcia — CHANGES_REQUESTED** [review]

File: portal-impl/src/com/liferay/portlet/documentlibrary/service/impl/DLFileEntryLocalServiceImpl.java [hunk 6]
```diff
@@ -366,6 +369,26 @@ public DLFileVersion cancelCheckOut(long userId, long fileEntryId)
...
 	}
 
+	@Override
+	public boolean canCopyFileEntry(
+		FileEntry fileEntry, long systemMaxSizeToCopy,
+		long companyMaxSizeToCopy, long groupMaxSizeToCopy) {
+
+		long dlFileEntrySize = fileEntry.getSize();
+
+		if (((dlFileEntrySize <= systemMaxSizeToCopy) ||
```
Comment (jaimelr10) [inline portal-impl/src/com/liferay/portlet/documentlibrary/service/impl/DLFileEntryLocalServiceImpl.java:379]: We can move this to

File: modules/apps/document-library/document-library-api/src/main/java/com/liferay/document/library/configuration/DLSizeLimitConfigurationProvider.java [hunk 7]
```diff
@@ -15,6 +15,8 @@
 @ProviderType
 public interface DLSizeLimitConfigurationProvider {
 
+	public boolean canCopy(long dlFolderSize, long groupId, long companyId);
```
Comment (AliciaGarciaGarcia) [inline modules/apps/document-library/document-library-api/src/main/java/com/liferay/document/library/configuration/DLSizeLimitConfigurationProvider.java:18]: change name of dLFolderSize, not always a folder.

Comment (AliciaGarciaGarcia) [inline modules/apps/document-library/document-library-api/src/main/java/com/liferay/document/library/configuration/DLSizeLimitConfigurationProvider.java:18]: normally 
method(long companyId, long groupId, _other_)

File: modules/apps/document-library/document-library-web/src/main/java/com/liferay/document/library/web/internal/portlet/action/CopyFolderMVCRenderCommand.java [hunk 8]
```diff
@@ -31,6 +49,59 @@
...
+
+		try {
+			DLFolder dlFolder = _dlFolderLocalService.getDLFolder(
+				ParamUtil.getLong(renderRequest, "sourceFolderId"));
+
+			long dlFolderSize = _dlFolderLocalService.getDLFolderSize(
+				dlFolder.getTreePath(), dlFolder.getCompanyId(),
+				dlFolder.getGroupId());
+
+			boolean canCopy = _dlSizeLimitConfigurationProvider.canCopy(
```
Comment (AliciaGarciaGarcia) [inline modules/apps/document-library/document-library-web/src/main/java/com/liferay/document/library/web/internal/portlet/action/CopyFolderMVCRenderCommand.java:65]: This method can be inlined or is SF change?

File: modules/apps/document-library/document-library-web/src/main/java/com/liferay/document/library/web/internal/portlet/action/CopyFolderMVCRenderCommand.java [hunk 9]
```diff
@@ -31,6 +49,59 @@
...
+				dlFolderSize, dlFolder.getGroupId(), dlFolder.getCompanyId());
+
+			if (!canCopy) {
```
Comment (AliciaGarciaGarcia) [inline modules/apps/document-library/document-library-web/src/main/java/com/liferay/document/library/web/internal/portlet/action/CopyFolderMVCRenderCommand.java:68]: IMO:
```
if (canCopy){
    return super.render(renderRequest, renderResponse);
}
```
...

File: modules/apps/document-library/document-library-web/src/main/java/com/liferay/document/library/web/internal/portlet/action/CopyFileEntryMVCRenderCommand.java [hunk 10]
```diff
@@ -31,6 +48,54 @@
...
 
+	@Override
+	public String render(
+			RenderRequest renderRequest, RenderResponse renderResponse)
+		throws PortletException {
+
+		try {
+			FileEntry fileEntry = ActionUtil.getFileEntry(renderRequest);
+
+			if (!_dlSizeLimitConfigurationProvider.canCopy(
```
Comment (AliciaGarciaGarcia) [inline modules/apps/document-library/document-library-web/src/main/java/com/liferay/document/library/web/internal/portlet/action/CopyFileEntryMVCRenderCommand.java:59]: change the if  to be the more readable
if (true){
return super...
}

(same comment than on the folder)

File: portal-kernel/src/com/liferay/document/library/kernel/service/DLFolderLocalServiceUtil.java [hunk 11]
```diff
@@ -484,6 +494,12 @@ public static int getDLFoldersCount() {
 		return getService().getDLFoldersCount();
 	}
 
+	public static long getDLFolderSize(
+		String treePath, long companyId, long groupId) {
```
Comment (AliciaGarciaGarcia) [inline portal-kernel/src/com/liferay/document/library/kernel/service/DLFolderLocalServiceUtil.java:498]: same as before
order:

long companyId, long groupId....

**Review by AliciaGarciaGarcia — CHANGES_REQUESTED** [review]

**Comment by AliciaGarciaGarcia** [comment]
Another thing...
The order of the commits is ... estrange.

buildWathever is made after the changes, usually :D

File: modules/apps/document-library/document-library-api/src/main/java/com/liferay/document/library/configuration/DLSizeLimitConfigurationProvider.java (hunk 7 above)
Comment (AliciaGarciaGarcia) [reply modules/apps/document-library/document-library-api/src/main/java/com/liferay/document/library/configuration/DLSizeLimitConfigurationProvider.java:18]: And...

I'm not sure about the name :S

File: portal-impl/src/com/liferay/portlet/documentlibrary/service/impl/DLFolderLocalServiceImpl.java [hunk 12]
```diff
@@ -331,6 +338,59 @@ public int getCompanyFoldersCount(long companyId) {
 		return dlFolderPersistence.countByCompanyId(companyId);
 	}
 
+	@Override
+	public Object[] getCopyFailInfo(
+		long dlFolderSize, long systemMaxSizeToCopy, long companyMaxSizeToCopy,
+		long groupMaxSizeToCopy, Portal portal, RenderRequest renderRequest) {
```
Comment (AliciaGarciaGarcia) [inline portal-impl/src/com/liferay/portlet/documentlibrary/service/impl/DLFolderLocalServiceImpl.java:344]: nor portal nor request

**Review by ambrinchaudhary — APPROVED** [review]
I added a minor question, there is no much to validate from the frontend side

File: modules/apps/document-library/document-library-api/src/main/java/com/liferay/document/library/configuration/DLSizeLimitConfigurationProvider.java (hunk 7 above)
Comment (jaimelr10) [reply modules/apps/document-library/document-library-api/src/main/java/com/liferay/document/library/configuration/DLSizeLimitConfigurationProvider.java:18]: We will go with `isCopyToAllowed`

File: modules/apps/document-library/document-library-web/src/main/java/com/liferay/document/library/web/internal/portlet/action/CopyFolderMVCRenderCommand.java (hunk 8 above)
Comment (jaimelr10) [reply modules/apps/document-library/document-library-web/src/main/java/com/liferay/document/library/web/internal/portlet/action/CopyFolderMVCRenderCommand.java:65]: I'll inline it :)

File: modules/apps/document-library/document-library-service/src/main/java/com/liferay/document/library/internal/configuration/DLSizeLimitConfigurationProviderImpl.java [hunk 13]
```diff
@@ -45,6 +47,28 @@ public Map<String, Long> getCompanyMimeTypeSizeLimit(long companyId) {
 			companyId);
 	}
 
+	@Override
+	public Object[] getCopyToFailInfo(
+		long companyId, long groupId, long size, Locale locale) {
+
+		long groupMaxSizeToCopy = getGroupMaxSizeToCopy(groupId);
```
Comment (AliciaGarciaGarcia) [inline modules/apps/document-library/document-library-service/src/main/java/com/liferay/document/library/internal/configuration/DLSizeLimitConfigurationProviderImpl.java:54]: I think that this should be:
long company
long group

..

you are using first the group so should be near the use. (I think see https://github.com/brianchandotcom/liferay-portal/commit/0d676d7823f15240543f5db2868a7cbc7e77a6a1)

File: modules/apps/document-library/document-library-service/src/main/java/com/liferay/document/library/internal/configuration/DLSizeLimitConfigurationProviderImpl.java [hunk 14]
```diff
@@ -45,6 +47,28 @@ public Map<String, Long> getCompanyMimeTypeSizeLimit(long companyId) {
 			companyId);
 	}
 
+	@Override
+	public Object[] getCopyToFailInfo(
```
Comment (AliciaGarciaGarcia) [inline modules/apps/document-library/document-library-service/src/main/java/com/liferay/document/library/internal/configuration/DLSizeLimitConfigurationProviderImpl.java:51]: I think that this can return an String[] ?

File: portal-impl/src/com/liferay/portlet/documentlibrary/service/impl/DLFolderLocalServiceImpl.java [hunk 15]
```diff
@@ -331,6 +334,35 @@ public int getCompanyFoldersCount(long companyId) {
 		return dlFolderPersistence.countByCompanyId(companyId);
 	}
 
+	@Override
+	public long getDLFolderSize(String treePath, long companyId, long groupId) {
```
Comment (AliciaGarciaGarcia) [inline portal-impl/src/com/liferay/portlet/documentlibrary/service/impl/DLFolderLocalServiceImpl.java:338]: we are on DLFolder file so there is no need to name the kind of folder

public long getFolderSize

see that on this file (DLFolderLocalService)
the names are 

getFile
getFolder
...

it's because we are on the "DL" kind of folder so the "DL" is implicit

**Review by AliciaGarciaGarcia — CHANGES_REQUESTED** [review]

**Comment by AliciaGarciaGarcia** [comment]
ci:test:sf

File: modules/apps/document-library/document-library-service/src/main/java/com/liferay/document/library/internal/configuration/DLSizeLimitConfigurationProviderImpl.java [hunk 16]
```diff
@@ -45,6 +47,28 @@ public Map<String, Long> getCompanyMimeTypeSizeLimit(long companyId) {
...
+
+		long companyMaxSizeToCopy = getCompanyMaxSizeToCopy(companyId);
+		long groupMaxSizeToCopy = getGroupMaxSizeToCopy(groupId);
+
+		if (((size > groupMaxSizeToCopy) && (groupMaxSizeToCopy != 0)) ||
+			((size > companyMaxSizeToCopy) && (companyMaxSizeToCopy != 0))) {
+
+			return new String[] {
+				_language.get(locale, "site-or-instance"),
+				_language.formatStorageSize(groupMaxSizeToCopy, locale)
```
Comment (AliciaGarciaGarcia) [inline modules/apps/document-library/document-library-service/src/main/java/com/liferay/document/library/internal/configuration/DLSizeLimitConfigurationProviderImpl.java:62]: This return the _groupMaxSizeToCopy_ always because will be the lesser or is a typo. 
Never will be the companyMaxSizeToCopy?

Comment (jaimelr10) [reply modules/apps/document-library/document-library-service/src/main/java/com/liferay/document/library/internal/configuration/DLSizeLimitConfigurationProviderImpl.java:62]: As we show on a call the other day, when we update the companyMaxSizeToCopy, the groupMaxSizeToCopy updates, but we can update groupMaxSizeToCopy without changing companyMaxSizeToCopy, so that's why

Comment (AliciaGarciaGarcia) [reply modules/apps/document-library/document-library-service/src/main/java/com/liferay/document/library/internal/configuration/DLSizeLimitConfigurationProviderImpl.java:62]: In my opinion, I think that is a good idea to split this behavior in to check both.
First:
is more accurate, if we change the behavior of the configuration saving, this won't has to be changed.
Second:
we are checking companyMaxSizeToCopy, you are checking it on the if but not using it.

But if you think is not worth to check it, for the behavior you mentioned before, maybe is better idea to remove from the if.

**Comment by jaimelr10** [comment]
ci:test:sf

**Comment by AliciaGarciaGarcia** [comment]
ci:test:sf

File: modules/apps/document-library/document-library-service/src/main/java/com/liferay/document/library/internal/configuration/DLSizeLimitConfigurationProviderImpl.java [hunk 17]
```diff
@@ -45,6 +47,33 @@ public Map<String, Long> getCompanyMimeTypeSizeLimit(long companyId) {
 			companyId);
 	}
 
+	@Override
+	public String[] getCopyToFailInfo(
+		long companyId, long groupId, long size, Locale locale) {
+
+		long groupMaxSizeToCopy = getGroupMaxSizeToCopy(groupId);
```
Comment (AliciaGarciaGarcia) [inline modules/apps/document-library/document-library-service/src/main/java/com/liferay/document/library/internal/configuration/DLSizeLimitConfigurationProviderImpl.java:54]: You are using this after the if. 
Declare it after then

**Review by AliciaGarciaGarcia — CHANGES_REQUESTED** [review]

**Comment by jaimelr10** [comment]
ci:test:sf

**Comment by AliciaGarciaGarcia** [comment]
ci:test:relevant

**Comment by Jeremy-Ch** [comment]
ci:test:relevant

**Comment by Jeremy-Ch** [comment]
The failures are unrelated.

**Comment by Jeremy-Ch** [comment]
ci:forward:force

**Comment by jaimelr10** [comment]
Resend with proposed comments to: https://github.com/liferay-lima/liferay-portal/pull/4619
